neqo-common/src/datagram.rs

```diff
@@ -7,20 +7,129 @@
 use std::net::SocketAddr;
 use std::ops::Deref;
 
+use enum_map::Enum;
+
 use crate::hex_with_len;
 
-#[derive(PartialEq, Eq, Clone)]
+// ECN (Explicit Congestion Notification) codepoints mapped to the
+// lower 2 bits of the TOS field.
+// https://www.iana.org/assignments/dscp-registry/dscp-registry.xhtml
+#[derive(Copy, Clone, PartialEq, Eq, Enum, Default)]
+#[repr(u8)]
+pub enum IpTosEcn {
+    #[default]
+    NotEct = 0b00, // Not-ECT, Not ECN-Capable Transport, [RFC3168]
+    Ect1 = 0b01, // ECT(1), ECN-Capable Transport(1), [RFC8311][RFC9331]
+    Ect0 = 0b10, // ECT(0), ECN-Capable Transport(0), [RFC3168]
+    Ce = 0b11,   // CE, Congestion Experienced, [RFC3168]
+}
+
+impl std::fmt::Debug for IpTosEcn {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            IpTosEcn::NotEct => f.write_str("Not-ECT"),
+            IpTosEcn::Ect1 => f.write_str("ECT(1)"),
+            IpTosEcn::Ect0 => f.write_str("ECT(0)"),
+            IpTosEcn::Ce => f.write_str("CE"),
+        }
+    }
+}
+
+// DiffServ Codepoints, mapped to the upper six bits of the TOS field.
+// https://www.iana.org/assignments/dscp-registry/dscp-registry.xhtml
+#[derive(Copy, Clone, PartialEq, Eq, Enum, Default)]
+#[repr(u8)]
+pub enum IpTosDscp {
+    #[default]
+    Cs0 = 0b0000_0000, // Class Selector 0, [RFC2474]
+    Cs1 = 0b0010_0000,        // Class Selector 1, [RFC2474]
+    Cs2 = 0b0100_0000,        // Class Selector 2, [RFC2474]
+    Cs3 = 0b0110_0000,        // Class Selector 3, [RFC2474]
+    Cs4 = 0b1000_0000,        // Class Selector 4, [RFC2474]
+    Cs5 = 0b1010_0000,        // Class Selector 5, [RFC2474]
+    Cs6 = 0b1100_0000,        // Class Selector 6, [RFC2474]
+    Cs7 = 0b1110_0000,        // Class Selector 7, [RFC2474]
+    Af11 = 0b0010_1000,       // Assured Forwarding 11, [RFC2597]
+    Af12 = 0b0011_0000,       // Assured Forwarding 12, [RFC2597]
+    Af13 = 0b0011_1000,       // Assured Forwarding 13, [RFC2597]
+    Af21 = 0b0100_1000,       // Assured Forwarding 21, [RFC2597]
+    Af22 = 0b0101_0000,       // Assured Forwarding 22, [RFC2597]
+    Af23 = 0b0101_1000,       // Assured Forwarding 23, [RFC2597]
+    Af31 = 0b0110_1000,       // Assured Forwarding 31, [RFC2597]
+    Af32 = 0b0111_0000,       // Assured Forwarding 32, [RFC2597]
+    Af33 = 0b0111_1000,       // Assured Forwarding 33, [RFC2597]
+    Af41 = 0b1000_1000,       // Assured Forwarding 41, [RFC2597]
+    Af42 = 0b1001_0000,       // Assured Forwarding 42, [RFC2597]
+    Af43 = 0b1001_1000,       // Assured Forwarding 43, [RFC2597]
+    Ef = 0b1011_1000,         // Expedited Forwarding, [RFC3246]
+    VoiceAdmit = 0b1011_0000, // Capacity-Admitted Traffic, [RFC5865]
+    Le = 0b0000_0100,         // Lower-Effort, [RFC8622]
+}
+
+impl std::fmt::Debug for IpTosDscp {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            IpTosDscp::Cs0 => f.write_str("CS0"),
```

Comment:
There's a little less typing if you do
```rust
let s = match self {
Self::foo -> "foo",
...
};
f.write_str(s)
```

Though I would say that it would also be OK to just `#[derive(Debug)]`.  It's just debugging.